src/Sensor.h

```diff
@@ -20,52 +20,101 @@
+		virtual void get_data(Vector3& data) = 0;
+		// virtual void get_data(double& data) = 0;
 
-protected:
-	Rocket& _rocket;
+		// Noise/bias injection control
+		void enable_noise_injection() 	{_inject_noise = true;};
+		void disable_noise_injection() 	{_inject_noise = false;};
+		void enable_bias_injection() 	{_inject_bias = true;};
+		void disable_bias_injection() 	{_inject_bias = false;};
 
-	double _refresh_rate;
-	double _last_update_tStep;
-	bool _new_data;
+	protected:
 
-	Vector3 _data;
-	double _data_scalar;
+		std::string _name;
+		
+		Rocket& _rocket;
+
+		double _refresh_rate;
+		double _last_update_tStep;
+		bool _new_data;
+
+		// Normal distribution noise generation
+		std::default_random_engine _generator;
+		std::normal_distribution<double> _normal_dist;
+		bool _inject_bias = false; 			// Whether to inject constant bias to measurement
+		bool _inject_noise = false; 		// Whether to inject random noise to measurement
 };
 
 class Gyroscope : public Sensor {
-public:
-	Gyroscope(Rocket& rocket, double refresh_rate);
-	void update_data(double tStep);
-	void get_data(Vector3& data);
+	public:
+		Gyroscope(std::string name, Rocket& rocket, double refresh_rate,
+				  double noise_mean=0.0f, double noise_stddev=0.1f);
+		void update_data(double tStep);
+		void get_data(Vector3& data);
+
+		void set_constant_bias(Vector3 bias) {_bias = bias;};
+
+	private:
+		Vector3 _data; 		// The sensor's current reading
+
+		Vector3 _noise; 	// Noise vector to be added to measurement
+				
+		Vector3 _bias; 		// Constant bias vector to be added to measurement	
 };
 
 class Accelerometer : public Sensor {
-public:
-	Accelerometer(Rocket& rocket, double refresh_rate);
-	void update_data(double tStep);
-	void get_data(Vector3& data);
+	public:
+		Accelerometer(std::string name, Rocket& rocket, double refresh_rate,
+					  double noise_mean=0.0f, double noise_stddev=0.1f);
+		void update_data(double tStep);
+		void get_data(Vector3& data);
+
+		void set_constant_bias(Vector3 bias) {_bias = bias;};
+
+ 	private:
+		Vector3 _data; 		// The sensor's current reading
+
+		Vector3 _noise; 	// Noise vector to be added to measurement
+
+		Vector3 _bias; 		// Constant bias vector to be added to measurement	
 };
 
+// TODO: Implement functions for both altitude and pressure measurements (and specify units)
 class Barometer : public Sensor {
-public:
-	Barometer(Rocket& rocket, double refresh_rate);
-	void update_data(double tStep);
-	void get_data(double& data);
+	public:
+		Barometer(std::string name, Rocket& rocket, double refresh_rate,
+				  double noise_mean=0.0f, double noise_stddev=0.1f);
+		void update_data(double tStep);
+		void get_data(double& data);
+
+		void set_constant_bias(double bias) {_bias = bias;};
+	
+	private:
+		double _data; 		// The sensor's current reading
+
+		double _noise; 		// Noise value to be added to measurement
+
+		double _bias; 		// Constant bias value to be added to measurement	
 };
```

Comment:
Maybe I'm missing some file but how are the accelerometer, barometer and gyro objects functionally different from each other?

Reply:
They're all very similar in that they get the ground truth values directly from the `Rocket` class and either inject noise/bias or directly output ground truth.

Difference is that `Accelerometer` takes the `r_ddot` (acceleration vector) from `Rocket` , whereas `Gyro` takes the `w_vect` (omega vector) from `Rocket`.

`Accelerometer` and `Gyro` also do coordinate frame transformations from inertial-frame to rocket-frame. Since `r_ddot` and `w_vect` are expressed in inertial-frame.

The barometer just takes the z axis scalar from the position vector of the rocket `r_vect`